scripts/lib/CIME/case/check_input_data.py

```diff
@@ -373,6 +373,7 @@ def check_input_data(case, protocol="svn", address=None, input_data_root=None, d
                             if chksum:
                                 verify_chksum(input_data_root, rundir, rel_path.strip(os.sep), isdirectory)
                                 logger.info("Chksum passed for file {}".format(os.path.join(input_data_root,rel_path)))
+                            expect((not isdirectory), "Unsupported directory in input_data_list '{}'".format(full_path))
```

Comment:
Can you explain the rationale for doing this check at this point? This will only do the check when the given file/directory already exists, as opposed to doing the isdirectory check regardless of existence. Perhaps more importantly, it looks like the other code around here is built to work with isdirectory true or false, so what is the rationale for adding this expectation that it is not a directory?

I don't understand the requirements here, so I'm asking these questions out of naivety.

Reply:
Hi Bill, the original issue was opened up by Benjamin Cash, and he was running into trouble with the check_input_data tool passing incorrect entries that pointed to existing directories, when they should have been pointing to existing files. Such as:

Buildconf/cice.input_data_list
grid_file = /work/02441/bcash/stampede2/UFS/ufs_inputdata/
kmt_file = /work/02441/bcash/stampede2/UFS/ufs_inputdata/

The subsequent run will fail, however, because grid_file and kmt_file are required.

I put the check there to fulfill his improvement request of having the check_input_data tool catch the type of errors described in his issue.

Reply:
I guess I'm questioning whether this response to the request is really the right one. It seems like this code has been designed to support directories as well as individual files, and so adding this assertion feels like it's changing this in a not-totally-consistent way. I feel like this needs more discussion. I'm also wondering how this change aligns with the discussion in #3654 , which showed that there are some cases where there are, indeed, directory paths in input data lists.

To summarize, I'm asking two things:

(1) I feel there needs to be broader discussion about whether check_input_data is intended to support directory paths or only individual files

(2) If it is only intended to support individual files, then – at a glance – it looks to me like your addition may be inconsistent with some existing code, and it may be that a bigger rework / cleanup is warranted. Along with this, the check should probably be moved to somewhere higher up in the conditional structure, so that it is checked whether or not a file already exists.